docs/guides/nextjs.md

```diff
@@ -0,0 +1,277 @@
+---
+title: Setup with Next.js
+nav: 21
+---
+
+[Next.js](https://nextjs.org) is a popular server-side rendering framework for React that presents
+some unique challenges for using Zustand properly.
+Keep in mind that Zustand store is a global
+variable (AKA module state) making it optional to use a `Context`.
+These challenges include:
+
+- **Per-request store:** A Next.js server can handle multiple requests simultaneously. This means
+  that the store should be created per request and should not be shared across requests.
+- **SSR friendly:** Next.js applications are rendered twice, first on the server
+  and again on the client. Having different outputs on both the client and the server will result
+  in "hydration errors." The store will have to be initialized on the server and then
+  re-initialized on the client with the same data in order to avoid that. Please read more about
+  that in our [SSR and Hydration](./ssr-and-hygration) guide.
+- **SPA routing friendly:** Next.js supports a hybrid model for client side routing, which means
+  that in order to reset a store, we need to intialize it at the component level using a
+  `Context`.
+- **Server caching friendly:** Recent versions of Next.js (specifically applications using the App
+  Router architecture) support aggressive server caching. Due to our store being a **module state**,
+  it is completely compatible with this caching.
+
+We have these general recommendations for the appropriate use of Zustand:
+
+- **No global stores** - Because the store should not be shared across requests, it should not be defined
+  as a global variable. Instead, the store should be created per request.
+- **React Server Components should not read from or write to the store** - RSCs cannot use hooks or context. They aren't
+  meant to be stateful. Having an RSC read from or write values to a global store violates the
+  architecture of Next.js.
+
+### Creating a store per request
+
+Let's write our store factory function that will create a new store for each
+request.
+
+```ts
+// stores/counter-store.ts
+import { createStore } from 'zustand'
+
+export type CounterState = {
+  count: number
+}
+
+export type CounterActions = {
+  decrementCount: () => void
+  incrementCount: () => void
+}
+
+export type CounterStore = CounterState & CounterActions
+
+export const defaultInitState: Partial<CounterState> = {
+  count: 0,
+}
+
+export const createCounterStore = (
+  initState: Partial<CounterState> = defaultInitState,
+) => {
+  return createStore<CounterStore>()((set) => ({
+    ...initState,
+    decrementCount: () => set((state) => ({ count: state.count - 1 })),
+    incrementCount: () => set((state) => ({ count: state.count + 1 })),
+  }))
+}
+```
+
+### Providing the store
+
+Let's use the `createCounterStore` in our component and share it using a context provider.
```

Comment:
Because `zustand/context` is deprecated, my team has been using my package [`zustand-di`](https://github.com/charkour/zustand-di) to avoid needing to re-write much of the boilerplate for using Zustand with React Context. If you think it will be helpful, you can update the guide to link to this package, but if not, we can leave this guide unchanged. Thank you

Reply:
For this guide I believe is good to have it as it is so people can understand what's under the hood